### PR TITLE
60 show notes of last x days

### LINKED ISSUE #60: Add option to show notes on last X days
For example: 

`stup show last`: Shows the last entry
`stup show last 2`: Show entries on the last 2 days **when entries exist**

### PR DESCRIPTION
**Pull request by ioanniswd**
Closes #60

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -265,7 +265,7 @@ $ stup log --from <from-date> --to <to-date> -c <category-name>
 
 # or using an alias
 
-$ stup log <week|previous-week|month|previous-month|year|previous-year>
+$ stup log <week|previous-week|days|month|previous-month|year|previous-year>
```
Comment (ghost) [inline README.md]: I don’t think this is correct – look how all the other variants make sense on their own, while `days` needs an argument.

File: README.md [hunk 2]
```diff
@@ -303,6 +303,17 @@ $ stup log previous-week
 $ stup previous-week
 ```
 
+##### Log notes of the last X days, where notes exist
+```bash
```
Comment (ghost) [inline README.md]: nitpick: missing empty line (the next section has it, haven’t checked the rest)

File: stup [hunk 3]
```diff
@@ -790,6 +795,18 @@ parse_options()
         fi
         shift 2
         ;;
+      --days|days)
+        if [ -z "$COMMAND" ]; then
+          COMMAND="log"
+        fi
+        if [ -z "$2" ]; then
+          DAYS=2
+          shift
```
Comment (ghost) [inline stup]: This doesn’t make much sense to me, at least the fallback should be configurable (IMO).

**Review by ghost — COMMENTED** [review]
I imagine it could be a wrapper for setting `--from` and `--to` for log, though that would require more thought (e.g. imagine `stup log --from 2021-10-02 --days 14`).

IIRC, there’s some difference between how `log` & `show` display the notes, so the internal usage of `show` also doesn’t seem appropriate to me.

But all this is just my own point of view, so take it with a pinch (or two) of salt.

File: README.md (hunk 1 above)
Comment (ioanniswd) [reply README.md]: a068731ecbf0830eb2a7949e5dd1190e5d15c686

File: README.md (hunk 2 above)
Comment (ioanniswd) [reply README.md]: c17257e8b4a63df5c1e4ca6b69728f9ccf3e14de

File: stup (hunk 3 above)
Comment (ioanniswd) [reply stup]: I agree.
ae4302465bc33e5a028e7b40b3c244ee5676f313

**Review by ioanniswd — COMMENTED** [review]
Thanks a lot for the review!

**Comment by ioanniswd** [comment]
@tiosgz 

> IIRC, there’s some difference between how log & show display the notes, so the internal usage of show also doesn’t seem appropriate to me.

You are not wrong about this. It would require some re-writing though. Thinking about it.

> I imagine it could be a wrapper for setting --from and --to for log, though that would require more thought (e.g. imagine stup log --from 2021-10-02 --days 14).

This option has a slightly different behavior than options `from` and `to`. It find the last X days where notes actually exist. I think it might be a little confusing when used with `from`/`to`

**Comment by ghost** [comment]
> This option has a slightly different behavior than options `from` and `to`. It find the last X days where notes actually exist. I think it might be a little confusing when used with `from`/`to`

It was just an idea, so again feel free to ignore this. But it doesn’t seem confusing to me:

- alone → last x days
- w/ from → x days beginning with `from` (because i may be lazy to count what date is x days after `from`)
- w/ to → x days before `to`

**Comment by ioanniswd** [comment]
@tiosgz I am convinced. Nice!
1111eae 
87736f7

File: stup [hunk 4]
```diff
@@ -501,6 +502,22 @@ execute_log()
     exit
   fi
 
+  if ! [ -z "$DAYS" ]; then
```
Comment (ghost) [inline stup]: `if ! [ -z "$var" ]` should be functionally the same as `if [ -n "$var" ]`

File: stup [hunk 5]
```diff
@@ -501,6 +502,22 @@ execute_log()
...
+      execute_show_last_x_days
+      exit
+    fi
+
+    if ! [ -z "$FROM_AT" ] && [ -z "$TO_AT" ]; then
+      TO_AT="$(date -d "$FROM_AT +$DAYS days" '+%F')"
+    elif ! [ -z "$TO_AT" ] && [ -z "$FROM_AT" ]; then
+      FROM_AT="$(date -d "$TO_AT $DAYS days ago" '+%F')"
+    else
+      print_error "You can only any pair of [\"--from\", \"--to\", \"--days\"], but not all three"
```
Comment (ghost) [inline stup]: You can only *what?* any pair… (it’s missing a verb)

**Review by ghost — COMMENTED** [review]
Oh, great! Last round of nitpicks, i hope 😆

(I must admit that i took the lazy way and didn’t previously inspect how the X was used, so now i see how it could be confusing. But now that i’ve convinced you, what’s the point? 😇)

File: stup (hunk 4 above)
Comment (ioanniswd) [reply stup]: 054abcb 
e10cb4f

File: stup (hunk 5 above)
Comment (ioanniswd) [reply stup]: 091a09a

**Review by ioanniswd — COMMENTED** [review]
:rocket:

**Comment by ioanniswd** [comment]
>IIRC, there’s some difference between how log & show display the notes, so the internal usage of show also doesn’t seem appropriate to me.

> > You are not wrong about this. It would require some re-writing though. Thinking about it.

I have a couple of ideas about this. Working on it.

Feel free to resolve any conversations if you find the fix ok.

**Comment by ioanniswd** [comment]
Looking at function `execute_log` more closely, I realized that I simply needed a way to calculate the dates with the notes I need to return. Will probably squash in a rebase, no reason to refactor in the same PR.
 a26f66c 

@tiosgz Let me know when you are ok with a rebase and a force push.

**Comment by ghost** [comment]
> Feel free to resolve any conversations if you find the fix ok.

Sorry, but the buttons aren’t there for me (idk how that happened). You can consider everything resolved as long as i don’t reply to it.

> Let me know when you are ok with a rebase and a force push.

Don’t let yourself get limited with this, the commits are still accessible (`force-pushed` contains a link to the force-push diff, the old & new commits link to… the commits).

File: README.md [hunk 6]
```diff
@@ -261,11 +261,13 @@ $ stup log previous-week
 The full version of the command:
 
 ```bash
-$ stup log --from <from-date> --to <to-date> -c <category-name>
+$ stup log --from <from-date> --to <to-date> -c <category-name> --days <days-before/after>
```
Comment (ghost) [inline README.md]: This suggests that all three can be used together, but unfortunately i don’t know how to make it clear. Perhaps `[ --from <from-date> | --to <to-date> | --days <days-before/after> ]`, but that’s debatable.

File: README.md [hunk 7]
```diff
@@ -261,11 +261,13 @@ $ stup log previous-week
...
 
 # or using an alias
 
 $ stup log <week|previous-week|month|previous-month|year|previous-year>
+
+$ stup log --days <number_of_days>
```
Comment (ghost) [inline README.md]: `--days` is already mentioned a few lines above, so this might be redundant.

File: README.md [hunk 8]
```diff
@@ -276,6 +278,7 @@ where:
   - this is also optional and if omitted the notes to be displayed won't have be **added before a specific date**
 * `-c` or `--category`: is the category option (optional). **If omitted, you will view the notes of all categories**
   * `<category-name>`: the name of the category whose notes you want to see
+* `days-before/after`: When used with option `--from`, it is treated as `<days-after>` and is used to calculate the `<to-date>` in relevance to the `<from-date>`, i.e. `<days-after>` after `<from-date>`. When used with option `<to-date>`, it is treated as `days-before` and is used to calculate the `from-date`, i.e. `<days-before>` before the `<to-date>`.
```
Comment (ghost) [inline README.md]: Seems too long to me 🤷. Perhaps “when used with `--from` or `--to`, denotes how many days after/before that date should be logged; when used alone, that many last days with notes are shown”?

File: stup [hunk 9]
```diff
@@ -501,27 +502,53 @@ execute_log()
...
+      echo -e "$(print_info "Displaying notes until") $(display_date "$TO_AT")"
+      target_command="sed "/$TO_AT/q""
+    else
+      # TODO: we should prompt for confirmation here: "Are you sure you want to view all of your notes?"
+      echo -e "$(print_info "Displaying all notes")"
+      target_command='tr -d ""'
+    fi
+
+    IFS=$'\n' read -d '' dates < <(
     echo -e "$FROM_AT\n$TO_AT\n$(find $REPOSITORY_ROOT -path "$REPOSITORY_ROOT/*/*/*/*.md")" |
```
Comment (ghost) [inline stup]: Forgot to indent?

**Review by ghost — COMMENTED** [review]
Oops, sorry to have let you wait. I ended up with a few comments even now, but i don’t know how to call them if the previous ones were nitpicks 😉

Well, one a little more serious thing is that the embedded help (`show_usage_log`) doesn’t mention `--days`.

File: README.md (hunk 6 above)
Comment (ioanniswd) [reply README.md]: I thought of the same at some point.
How about this:
2593e7de5aac56322fef6434ac6fb7c542676069

File: README.md (hunk 7 above)
Comment (ioanniswd) [reply README.md]: Right you are
e3eacd2b25487b7f0d6d6b92e55f7afa8cb306ee

File: README.md (hunk 8 above)
Comment (ioanniswd) [reply README.md]: It is a bit long.
92433cc964428d6360ba852f334b3e0fa0381354

File: stup (hunk 9 above)
Comment (ioanniswd) [reply stup]: :mag: 
Copy paste shenanigan
bf3ab03

**Review by ioanniswd — COMMENTED** [review]
Thank you for being so thorough! :mag:

**Review by ioanniswd — COMMENTED** [review]
> Well, one a little more serious thing is that the embedded help (show_usage_log) doesn’t mention --days.

Missed it :)
cb209f7

File: README.md (hunk 6 above)
Comment (ghost) [reply README.md]: Grouping the three time-limiting args (so that `-c` is first / last) would be a better option. Otherwise looks good.

Comment (ioanniswd) [reply README.md]: You mean like this:
```
$ stup log [--from <from-date>] [--to <to-date>] [--days <days-before/after>] [-c <category-name>] 
```

File: stup [hunk 10]
```diff
@@ -1620,6 +1620,9 @@ List of available options:"
   printf "\n%15s   %s" " "               "Examples:"
   printf "\n%15s   %s" " "               "  stup log --from today --to 2020-08-10"
   printf "\n%15s   %s" " "               "  stup log --from 2019-04-18 --to today"
+  printf "\n%15s   %s" "--days, days"    "Shows the notes of the last <days> days."
```
Comment (ghost) [inline stup]: Last \<days\> days **with notes**.

Also, just realised how weird `stup log days 3` looks (compared to `stup log --days 3`). Is that an intentional thing, or were you confused by the args that don’t require `--`?

Comment (ioanniswd) [reply stup]: 867befcef49c7a7f20af89c1be6db63576023926

It was intentional, but you have a point 
71f42a0 
794b9b3

File: README.md (hunk 6 above)
Comment (ghost) [reply README.md]: Yup, just reorder it.

Comment (ioanniswd) [reply README.md]: 7c3642e

**Review by ghost — APPROVED** [review]
LGTM, thanks for your patience with me!

**Comment by ioanniswd** [comment]
Thank you for your help! :)

**Pull request merged**